### PR TITLE
Skip element if errors in metadata server

### PR DESCRIPTION
**Pull request by outscale-mdr**
Right now if some elements are not available in the metadata, the retrieval fails and it crashs. We propose to add the ability to skip some element if necessary.

Use-case: our tags can have invalid URI characters like '/' in the middle. Wewould like to be able to continue if some elements are not available with something like:

```python
def _skip_tags(exception):
    if isinstance(exception, uhelp.UrlError) and exception.code == 404:
        if "meta-data/tags/" in exception.url:
            print(exception.url)
            return True
    return False
data = ec2.get_instance_metadata(
                headers_cb=None,
                skip_cb=_skip_tags
            )
```

### CODE REVIEW

**Comment by outscale-mdr** [comment]
Add Outscale as datasource:

- Outscale machine are identified with a system-uuid starting with `osc`
- We skip errors when we retrieve metdata only when it is inside `tags` folder

**Comment by outscale-mdr** [comment]
TODO List (from [official website](https://cloudinit.readthedocs.io/en/latest/topics/datasources.html)):
- [X] Create a DataSource
- [X] Add tests for the datasource in (`cloudinit/sources/test_<yourplatform>.py`)
- [X] Update ds-identify 
- [X] Add tests of ds-identify
- [X] Update `cloudinit/settings.py`
- [X] Update `cloudinit/apport.py`
- [ ] Update `debian/cloud-init.templates`
- [X] Add Documentation in `doc/datasources/<cloudplatform>.rst`

File: tests/unittests/sources/test_outscale.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        ret = osc._is_outscale()
+        self.assertEqual(
+            self.read_dmi_data_expected, m_read_dmi_data.call_args_list
+        )
+        self.assertEqual(False, ret)
+
+    @mock.patch("cloudinit.sources.DataSourceOutscale.dmi.read_dmi_data")
+    def test_false_on_unknown_string(self, m_read_dmi_data):
+        """Should return false on an unrelated string."""
+        m_read_dmi_data.return_value = "pas osc dedans"
```
Comment (jerome-jutteau) [inline tests/unittests/sources/test_outscale.py]: no osc here

File: tests/unittests/sources/test_outscale.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        self.assertEqual(False, ret)
+
+
+class TestIsOutscale(test_helpers.CiTestCase):
+    read_dmi_data_expected = [mock.call("system-uuid")]
+
+    @mock.patch("cloudinit.sources.DataSourceOutscale.dmi.read_dmi_data")
+    def test_true_on_outscale_product(self, m_read_dmi_data):
+        """Should return true if the dmi product data has expected value."""
+        m_read_dmi_data.return_value = "oscetlasuite"
```
Comment (jerome-jutteau) [inline tests/unittests/sources/test_outscale.py]: osc and other things

File: doc/rtd/topics/datasources/outscale.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
+.. _datasource_outscale:
+
+3DS Outscale (Outscale)
+======================
+The ``Outscale`` datasource reads data from 3DS Outscale.  Support is
+present in cloud-init since 0.7.9.
```
Comment (jerome-jutteau) [inline doc/rtd/topics/datasources/outscale.rst]: "is include" ?

**Review by jerome-jutteau — CHANGES_REQUESTED** [review]

**Pull request closed without merging**